packages/aws-cdk-lib/aws-lambda/lib/function.ts

```diff
@@ -671,7 +671,18 @@ export class Function extends FunctionBase {
    * in the same account and region as the stack you are importing it into.
    */
   public static fromFunctionArn(scope: Construct, id: string, functionArn: string): IFunction {
-    return Function.fromFunctionAttributes(scope, id, { functionArn });
+    /**
+     * First, validate the functionArn string. If it's ARN with trailing version or alias,
+     * we trim off the trailing to retrieve the real functionArn.
+     * See lambda resource ARN format here: https://docs.aws.amazon.com/lambda/latest/dg/lambda-api-permissions-ref.html
+     */
+    const parts = functionArn.split(':');
+    if (parts.length > 7) {
```

Comment:
I'm not really a fan hardcoding this check with `> 7` but I also don't really foresee this ever changing in terms of sections of an ARN so I guess it's ok.

Reply:
@paulhcsun see my comment [here](https://github.com/aws/aws-cdk/commit/819888470248fed2ddbdac689f14254688eef88f#r140934472) too.  It is about the fact that the functionArn can be a token instead of a literal string too (for example when it is passed as a parameter to the CloudFormation stack). I also added a code snippet on how to handle that case too.

Reply:
fromFunctionArn only accepts functionArn as `string`.  Should we consider that condition?

Reply:
@pahud and @paulhcsun in the following code snippet is a stack that accepts the lambda function version arn as a parameter (for example because the actual function was created in another stack earlier in the ci/cd pipeline). 

```
const myFunctionVersionArn = new cdk.CfnParameter(stack, 'MyFunctionVersionArn');

const version = lambda.Version.fromVersionArn(stack, 'MyVersionFunction', myFunctionVersionArn.valueAsString);
```

The arn that is passed to fromVersionArn is a tokenized string (value not known at compile time) and is eventually passed to fromFunctionArn because it is called by fromVersionArn under the hood. And fromVersionArn passes it down without stripping the version part in the arn.

At the moment the fix has been applied to Function.fromFunctionArn, but I think you may assume people pass the correct arn to that function. The real problem is in fromVersionArn. This function "knows" it will receive a version arn so it also should know that it should strip the version to get a valid functionArn. So I think [this line](https://github.com/aws/aws-cdk/blob/5a918d12ec0455626b5630b35aa47e6d9da6df7c/packages/aws-cdk-lib/aws-lambda/lib/lambda-version.ts#L125) should be fixed. If you do that then the `Fn.join(Fn.select(.., Fn.split...)` will only be generated when it is really needed.

```
public static fromVersionArn(scope: Construct, id: string, versionArn: string): IVersion {
      const version = extractQualifierFromArn(versionArn);

      let functionArn: string;

      if (Token.isUnresolved(versionArn)) {
            const parts =  Fn.split(':', versionArn);
            functionArn = cdk.Fn.join(':', [
                Fn.select(0, parts),
                Fn.select(1, parts),
                Fn.select(2, parts),
                Fn.select(3, parts),
                Fn.select(4, parts),
                Fn.select(5, parts),
                Fn.select(6, parts),
            ]);
      } else {
            const parts = versionArn.split(':');
            functionArn = parts.slice(0, 7).join(':');
      }

      const lambda = Function.fromFunctionArn(scope, `${id}Function`, functionArn);
  
      class Import extends QualifiedFunctionBase implements IVersion {
        public readonly version = version;
        public readonly lambda = lambda;
        public readonly functionName = `${lambda.functionName}:${version}`;
        public readonly functionArn = versionArn;
        public readonly grantPrincipal = lambda.grantPrincipal;
        public readonly role = lambda.role;
        public readonly architecture = lambda.architecture;
  
        protected readonly qualifier = version;
        protected readonly canCreatePermissions = this._isStackAccount();
  
        public addAlias(name: string, opts: AliasOptions = {}): Alias {
          return addAlias(this, this, name, opts);
        }
  
        public get edgeArn(): string {
          if (version === '$LATEST') {
            throw new Error('$LATEST function version cannot be used for Lambda@Edge');
          }
          return this.functionArn;
        }
      }
      return new Import(scope, id);
    }
```